### PR TITLE
[numpy] add op linalg solve 

### PR DESCRIPTION
**Pull request by sjtuWangDing**
## Description ##
add op linalg solve 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/ndarray/numpy/linalg.py [hunk 1]
```diff
@@ -352,3 +352,58 @@ def slogdet(a):
...
+
+
+def solve(a, b):
+    r"""
+    Solve a linear matrix equation, or system of linear scalar equations.
+
+    Computes the "exact" solution, x, of the well-determined, i.e., full rank, linear matrix equation ax = b.
+
+    Parameters :
+    a : (..., M, M) ndarray
```
Comment (reminisce) [inline python/mxnet/ndarray/numpy/linalg.py]: Indent. And same for all the following doc. Please follow the same format as in

https://github.com/numpy/numpy/blob/0725a58415e68eab9fa7d7389cdb40716447a4fb/numpy/linalg/linalg.py#L326-L381

File: src/operator/numpy/linalg/np_solve-inl.h [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+                     const Tensor<xpu, 1, int>& ipiv,
+                     Stream<xpu> *s);
+
+template<typename xpu, typename DType> inline
+void linalg_dn_getrs(const Tensor<xpu, 2, DType>& A,
+                     const Tensor<xpu, 2, DType>& X,
+                     const Tensor<xpu, 1, int>& ipiv,
+                     Stream<xpu> *s);
+
+inline bool SolveOpShape(const nnvm::NodeAttrs &attrs,
```
Comment (reminisce) [inline src/operator/numpy/linalg/np_solve-inl.h]: Put shape and dtype inference functions in .cc.

File: python/mxnet/ndarray/numpy/linalg.py [hunk 3]
```diff
@@ -352,3 +352,57 @@ def slogdet(a):
...
+        Ordinate or "dependent variable" values.
+
+    Returns
+    -------
+    x : {(..., M,), (..., M, K)} ndarray
+        Solution to the system a x = b.  Returned shape is identical to `b`.
+
+    Raises
+    ------
+    LinAlgError
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/linalg.py]: Are we really raising this same error when input is singular or not square?

Comment (sjtuWangDing) [reply python/mxnet/ndarray/numpy/linalg.py]: If the input is singular, cblas function will raise error. If the input is not square, shape inference will raise error

Comment (haojin2) [reply python/mxnet/ndarray/numpy/linalg.py]: I mean are we raising an error of type `MXNetError` or `LinAlgError`?

Comment (sjtuWangDing) [reply python/mxnet/ndarray/numpy/linalg.py]: MXNetError

Comment (haojin2) [reply python/mxnet/ndarray/numpy/linalg.py]: So you need to change the doc to reflect that

Comment (sjtuWangDing) [reply python/mxnet/ndarray/numpy/linalg.py]: Ok

File: python/mxnet/numpy/linalg.py [hunk 4]
```diff
@@ -370,3 +370,56 @@ def slogdet(a):
     (1., -1151.2925464970228)
     """
     return _mx_nd_np.linalg.slogdet(a)
+
```
Comment (haojin2) [inline python/mxnet/numpy/linalg.py]: one more blank line.

File: src/operator/numpy/linalg/np_solve-inl.h [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+                     Stream<xpu> *s);
+
+template<typename xpu, typename DType> inline
+void linalg_dn_getrs(const Tensor<xpu, 2, DType>& A,
+                     const Tensor<xpu, 2, DType>& X,
+                     const Tensor<xpu, 1, int>& ipiv,
+                     Stream<xpu> *s);
+
+// kernel for conversing type
+struct SolveTypeHelper {
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_solve-inl.h]: there's a kernel named `identity_with_cast` in src/operator/mshadow_op.h

File: src/operator/numpy/linalg/np_solve-inl.h [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+    mxnet::TShape b_shape(a_shape.ndim(), 1);
+    for (int i = 0; i < a_shape.ndim() - 1; ++i) { b_shape[i] = b_tblob.shape_[i]; }
+    if (b_tblob.shape_.ndim() == a_shape.ndim()) {
+      b_shape[a_shape.ndim() - 1] = b_tblob.shape_[a_shape.ndim() - 1];
+    }
+    const int ndim = a_shape.ndim();
+    mxnet::TShape ipiv_shape(a_shape);
+    ipiv_shape[ndim - 1] = 1;
+    if (0 == a_shape[ndim - 1] || 0 == a_shape[ndim - 2]
+      || 0 == b_shape[ndim - 1] || 0 == b_shape[ndim - 2]) { return; }
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_solve-inl.h]:
```c++
    if (a_shape[ndim - 1] == 0 || a_shape[ndim - 2] == 0 ||
        b_shape[ndim - 1] == 0 || b_shape[ndim - 2] == 0) {
      return;
    }
```

File: src/operator/numpy/linalg/np_solve-inl.h [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+    const mxnet::TShape& a_shape = a_tblob.shape_;
+    mxnet::TShape b_shape(a_shape.ndim(), 1);
+    for (int i = 0; i < a_shape.ndim() - 1; ++i) { b_shape[i] = b_tblob.shape_[i]; }
+    if (b_tblob.shape_.ndim() == a_shape.ndim()) {
+      b_shape[a_shape.ndim() - 1] = b_tblob.shape_[a_shape.ndim() - 1];
+    }
+    const int ndim = a_shape.ndim();
+    const int N = a_shape[ndim - 1];
+    if (0 == a_shape[ndim - 1] || 0 == a_shape[ndim - 2]
+      || 0 == b_shape[ndim - 1] || 0 == b_shape[ndim - 2]) { return; }
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_solve-inl.h]: same here

File: src/operator/numpy/linalg/np_solve.cc [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+.set_attr<FCompute>("FCompute<cpu>", LaOpForwardSolve<cpu, 2, 2, 2, 1, solve>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{"_backward_npi_solve"})
+.add_argument("A", "NDArray-or-Symbol", "Tensor of square matrix")
+.add_argument("B", "NDArray-or-Symbol", "Tensor of right side vector");
+
+NNVM_REGISTER_OP(_backward_npi_solve)
+.set_num_inputs(4)
+.set_num_outputs(2)
+.set_attr<FResourceRequest>("FResourceRequest", [](const NodeAttrs& ){
+  return std::vector<ResourceRequest>{ResourceRequest::kTempSpace}; })
```
Comment (haojin2) [inline src/operator/numpy/linalg/np_solve.cc]:
```c++
.set_attr<FResourceRequest>("FResourceRequest", [](const NodeAttrs& ){
  return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
})
```

File: tests/python/unittest/test_numpy_op.py [hunk 9]
```diff
@@ -3412,6 +3412,97 @@ def check_inv(A_inv, data_np):
...
+
+    def check_solve(x, a_np, b_np):
+        try:
+            x_expected = _np.linalg.solve(a_np, b_np)
+        except Exception as e:
+            print(a_np)
+            print(a_np.shape)
+            print(b_np)
+            print(b_np.shape)
+            print(e)
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: get a better error message

**Pull request merged**